subsys/can/isotp/isotp.c

```diff
@@ -0,0 +1,1247 @@
+/*
+ * Copyright (c) 2019 Alexander Wachter
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "isotp_internal.h"
+#include <net/buf.h>
+#include <kernel.h>
+#include <init.h>
+#include <misc/util.h>
+
+#define LOG_LEVEL CONFIG_CAN_ISOTP_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(isotp);
+
+#ifdef CONFIG_CAN_ISOTP_ENABLE_CONTEXT_BUFFERS
+K_MEM_SLAB_DEFINE(ctx_slab, sizeof(struct iso_tp_send_ctx),
+		  CONFIG_CAN_ISOTP_TX_CONTEXT_BUF_COUNT, 4);
+#endif
+
+static void receive_pool_free(struct net_buf *buf);
+static void receive_ff_sf_pool_free(struct net_buf *buf);
+
+NET_BUF_POOL_DEFINE(isotp_rx_pool, CONFIG_CAN_ISOTP_RX_BUF_COUNT,
+		    CONFIG_CAN_ISOTP_RX_BUF_SIZE, sizeof(u32_t),
+		    receive_pool_free);
+
+NET_BUF_POOL_DEFINE(isotp_rx_sf_ff_pool, CONFIG_CAN_ISOTP_RX_SF_FF_BUF_COUNT,
+		    ISO_TP_CAN_DL, sizeof(u32_t), receive_ff_sf_pool_free);
+
+static struct iso_tp_global_ctx global_ctx = {
+	.alloc_list = SYS_SLIST_STATIC_INIT(&global_ctx.alloc_list),
+	.ff_sf_alloc_list = SYS_SLIST_STATIC_INIT(&global_ctx.ff_sf_alloc_list)
+};
+
+#ifdef CONFIG_CAN_ISOTP_USE_TX_BUF
+NET_BUF_POOL_VAR_DEFINE(isotp_tx_pool, CONFIG_CAN_ISOTP_TX_BUF_COUNT,
+			CONFIG_CAN_ISOTP_BUF_TX_DATA_POOL_SIZE, NULL);
+#endif
+
+K_THREAD_STACK_DEFINE(tx_stack, 512);
```

Comment:
The stack size must be settable from kconfig file. Some of the testing options, like enabling coverage generation, requires much more stack space and it is annoying to change the size if we have hard coded it like this.

Reply:
Is there a good way to check the real demand on the stack? I didn't use any recursive calls and avoided things that could increase the stack unexpected.

Reply:
> Is there a good way to check the real demand on the stack?

Not that I know of. The stack size is typically adjusted after getting some experience with the system.

Reply:
My experience told me to use 256 now :-). I checked with stack sentinel and assertions enabled.
192 crashes, 256 is good.